### ISSUE TITLE
Run into internal compiler error

### ISSUE THREAD

**Issue by fengsp** [rust-lang/rust#22799]
I am trying to compile https://github.com/fengsp/pencil, and I get following error:

    src/pencil/lib.rs:1:1: 1:1 error: internal compiler error: Encountered errors `[FulfillmentError(Obligation(predicate=Binder(TraitPredicate(core::clone::Clone)),depth=1),Ambiguity)]` fulfilling during trans
    src/pencil/lib.rs:1 // A microframework for Rust.
                        ^
    note: the compiler unexpectedly panicked. this is a bug.
    note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
    note: run with `RUST_BACKTRACE=1` for a backtrace
    thread 'rustc' panicked at 'Box<Any>', /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/libsyntax/diagnostic.rs:129

    stack backtrace:
       1:        0x112c14be3 - sys::backtrace::write::h918207864656307b0nC
       2:        0x112c42be5 - panicking::on_panic::h2c4b1456225ae37fnPL
       3:        0x112b6c418 - rt::unwind::begin_unwind_inner::h6c7f1344c68dd4a7AwL
       4:        0x1122226cf - rt::unwind::begin_unwind::h14266089743717698327
       5:        0x11222267c - diagnostic::SpanHandler::span_bug::hcfbef1e15a154a8dqZE
       6:        0x10f8ce10d - session::Session::span_bug::hed91eb24b076ccd3ynt
       7:        0x10f15b58d - trans::common::fulfill_obligation::h46ed86aa9b7a3242BGl
       8:        0x10f0f65d6 - trans::meth::trans_static_method_callee::h2a47652ca26cacacmjz
       9:        0x10f0f355b - trans::callee::trans::h1ce2921bf10326a8Kgg
      10:        0x10f10990b - trans::callee::trans_call_inner::h5306915459861161960
      11:        0x10f115daa - trans::expr::trans_rvalue_dps_unadjusted::hb1406d54d45518409Wi
      12:        0x10f0c69eb - trans::expr::trans_into::h0f31dfb1ff207cd8pGh
      13:        0x10f149f28 - trans::expr::trans_adt::hf2e69cd78868bafdzvj
      14:        0x10f14c401 - trans::expr::trans_struct::closure.42750
      15:        0x10f13339c - trans::expr::trans_struct::hf173464044f89757frj
      16:        0x10f1157d2 - trans::expr::trans_rvalue_dps_unadjusted::hb1406d54d45518409Wi
      17:        0x10f0c69eb - trans::expr::trans_into::h0f31dfb1ff207cd8pGh
      18:        0x10f1bbfd7 - trans::_match::trans_match_inner::h56e2641ee60bee0daMx
      19:        0x10f1151b3 - trans::expr::trans_rvalue_dps_unadjusted::hb1406d54d45518409Wi
      20:        0x10f0c69eb - trans::expr::trans_into::h0f31dfb1ff207cd8pGh
      21:        0x10f0c77fa - trans::controlflow::trans_block::hb675814db44e1ef35ee
      22:        0x10f19e2bf - trans::base::trans_closure::h1fc9dae5b42c150ephu
      23:        0x10f0b2086 - trans::base::trans_fn::h1baa1a79c6642379Gsu
      24:        0x10f0adaa4 - trans::base::trans_item::h99ee5fcee3ea0222zRu
      25:        0x10f19fa89 - trans::base::trans_mod::h0f4f7f5d04ce3f65yXu
      26:        0x10f0ad0dd - trans::base::trans_item::h99ee5fcee3ea0222zRu
      27:        0x10f1a4e0c - trans::base::trans_crate::h89e3520c5dfd05b21Ov
      28:        0x10ef31058 - driver::phase_4_translate_to_llvm::h5a7c1f8c7ec04632qPa
      29:        0x10ef0bf72 - driver::compile_input::h32723f720e0c3c95Gba
      30:        0x10efe48a7 - run_compiler::hfdf8469da778a8d4Zbc
      31:        0x10efe1d21 - thunk::F.Invoke<A, R>::invoke::h56535309890628509
      32:        0x10efe0980 - rt::unwind::try::try_fn::h6830563953582217021
      33:        0x112cbc7e9 - rust_try_inner
      34:        0x112cbc7d6 - rust_try
      35:        0x10efe10e5 - thunk::F.Invoke<A, R>::invoke::h12775416177659868106
      36:        0x112c2bac3 - sys::thread::thread_start::hccc0587963a6c50bU4G
      37:     0x7fff9a257268 - _pthread_body
      38:     0x7fff9a2571e5 - _pthread_body

    Could not compile `pencil`.

rustc --version --verbose:

    rustc 1.0.0-nightly (2b01a37ec 2015-02-21) (built 2015-02-21)
    binary: rustc
    commit-hash: 2b01a37ec38db9301239f0c0abcf3c695055b0ff
    commit-date: 2015-02-21
    build-date: 2015-02-21
    host: x86_64-apple-darwin
    release: 1.0.0-nightly

**Comment by gkoz** [comment]
I'm getting a very similar ICE with this reduced sample:
```rust
#![feature(core)]
use std::marker::{MarkerTrait, PhantomFn};

pub trait Wrapper {
    type Inner;
}

pub trait IsA<T>: MarkerTrait + PhantomFn<T> { }

impl <T> IsA<T> for T { }

pub trait ToPtr<P: Copy> {
    fn borrow(&self) -> P;
}

impl <PAR: Wrapper, CHI: IsA<PAR>> ToPtr<*mut <PAR as Wrapper>::Inner> for CHI {
    fn borrow(&self) -> *mut <PAR as Wrapper>::Inner {
        unimplemented!()
    }
}

pub struct Object;
pub struct Window;

impl Wrapper for Window {
    type Inner = i8;
}

impl IsA<Object> for Window { }

impl Window {
    pub fn new() -> Window {
        unimplemented!()
    }
}

pub trait WindowTrait {
    fn bar(&self);
}

impl <T: Wrapper> WindowTrait for T {
    fn bar(&self) {
        ToPtr::<_>::borrow(self);
    }
}

fn main() {
    let window = Window::new();
    window.bar();
}
```
The error message:
```
lib.rs:1:1: 1:1 error: internal compiler error: Encountered errors `[FulfillmentError(Obligation(predicate=Binder(TraitPredicate(core::marker::Sized)),depth=1),Ambiguity), FulfillmentError(Obligation(predicate=Binder(TraitPredicate(Wrapper)),depth=1),Ambiguity), FulfillmentError(Obligation(predicate=Binder(TraitPredicate(IsA<_>)),depth=1),Ambiguity), FulfillmentError(Obligation(predicate=Binder(ProjectionPredicate(<_ as Wrapper>::Inner, i8)),depth=2),Ambiguity)]` fulfilling during trans
lib.rs:1 #![feature(core)]
         ^
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libsyntax/diagnostic.rs:129

stack backtrace:
   1:     0x7f76c6609e2f - sys::backtrace::write::h947ec7660f7e4c8cSBC
   2:     0x7f76c662ff82 - panicking::on_panic::h23ca8edeaea6c81cLMI
   3:     0x7f76c6564e8a - rt::unwind::begin_unwind_inner::h47bdb490e575cf7eRsI
   4:     0x7f76c39db94d - rt::unwind::begin_unwind::h18314895764715619195
   5:     0x7f76c39db8f3 - diagnostic::SpanHandler::span_bug::h8ef69e7b1a01e686EaB
   6:     0x7f76c4294263 - session::Session::span_bug::h45f93f11f39024fbNsn
   7:     0x7f76c5cf5cc7 - trans::common::fulfill_obligation::h1573f2a041dd4f61ykl
   8:     0x7f76c5c9175b - trans::meth::trans_static_method_callee::hd5b7480b401c3e4aA7x
   9:     0x7f76c5c8e200 - trans::callee::trans::hb832566848979f51eYf
  10:     0x7f76c5ca4b7b - trans::callee::trans_call_inner::h13086090426327833152
  11:     0x7f76c5cb1820 - trans::expr::trans_rvalue_dps_unadjusted::h8c6df01d48428517xzi
  12:     0x7f76c5c61cf0 - trans::expr::trans_into::h9649503c823db95eSmh
  13:     0x7f76c5c60fe4 - trans::controlflow::trans_stmt_semi::hc954ae953b072232G5d
  14:     0x7f76c5c62ab0 - trans::controlflow::trans_block::h4c660619c3bd7816t6d
  15:     0x7f76c5d3a611 - trans::base::trans_closure::hcfabd7fd269607bekjt
  16:     0x7f76c5c4b238 - trans::base::trans_fn::hc85106bcc218fcb4dut
  17:     0x7f76c5c4c90b - trans::monomorphize::monomorphic_fn::h22691e7b0d2d168c2td
  18:     0x7f76c5c92a0e - trans::callee::trans_fn_ref_with_substs::h46ce93a241a2e351nkg
  19:     0x7f76c5caaa75 - trans::meth::trans_method_callee::h2c272c3ddb1fcd8cr2x
  20:     0x7f76c5ca64a9 - trans::callee::trans_call_inner::h320915813152599062
  21:     0x7f76c5cb05c9 - trans::expr::trans_rvalue_dps_unadjusted::h8c6df01d48428517xzi
  22:     0x7f76c5c61cf0 - trans::expr::trans_into::h9649503c823db95eSmh
  23:     0x7f76c5c60fe4 - trans::controlflow::trans_stmt_semi::hc954ae953b072232G5d
  24:     0x7f76c5c62ab0 - trans::controlflow::trans_block::h4c660619c3bd7816t6d
  25:     0x7f76c5d3a611 - trans::base::trans_closure::hcfabd7fd269607bekjt
  26:     0x7f76c5c4b238 - trans::base::trans_fn::hc85106bcc218fcb4dut
  27:     0x7f76c5c46ef1 - trans::base::trans_item::h167dcb48771cb88c5Rt
  28:     0x7f76c5d4373c - trans::base::trans_crate::h947378b1d60dae25wOu
  29:     0x7f76c6c3c6e3 - driver::phase_4_translate_to_llvm::h8e4d85cc4e2e15a5gOa
  30:     0x7f76c6c182d3 - driver::compile_input::hf1ccb6f79c82b5e3Nba
  31:     0x7f76c6cd3827 - run_compiler::h5744096088d771cbu2b
  32:     0x7f76c6cd1459 - thunk::F.Invoke<A, R>::invoke::h2151708636338439058
  33:     0x7f76c6cd0110 - rt::unwind::try::try_fn::h17480117432322948657
  34:     0x7f76c66a19c8 - rust_try_inner
  35:     0x7f76c66a19b5 - rust_try
  36:     0x7f76c6cd088a - thunk::F.Invoke<A, R>::invoke::h4030886730675850734
  37:     0x7f76c661e9a5 - sys::thread::thread_start::he7ae8e81d7f8a1f4x4G
  38:     0x7f76c04bbee4 - start_thread
  39:     0x7f76c61d7d1c - clone
  40:                0x0 - <unknown>
```

**Comment by frewsxcv** [comment]
In order to compile @gkoz's code block, I removed the `#![feature(core)]` line

That said, this still results in an ICE using Rust 1.0 beta 2

```
~/Downloads $ rustc test.rs
test.rs:7:19: 7:30 warning: use of deprecated item: No longer needed, #[warn(deprecated)] on by default
test.rs:7 pub trait IsA<T>: MarkerTrait + PhantomFn<T> { }
                            ^~~~~~~~~~~
test.rs:7:33: 7:45 warning: use of deprecated item: No longer needed, #[warn(deprecated)] on by default
test.rs:7 pub trait IsA<T>: MarkerTrait + PhantomFn<T> { }
                                          ^~~~~~~~~~~~
test.rs:1:1: 1:1 error: internal compiler error: Encountered errors `[FulfillmentError(Obligation(predicate=Binder(TraitPredicate(core::marker::Sized)),depth=1),Ambiguity), FulfillmentError(Obligation(predicate=Binder(TraitPredicate(Wrapper)),depth=1),Ambiguity), FulfillmentError(Obligation(predicate=Binder(TraitPredicate(IsA<_>)),depth=1),Ambiguity), FulfillmentError(Obligation(predicate=Binder(ProjectionPredicate(<_ as Wrapper>::Inner, i8)),depth=2),Ambiguity)]` fulfilling during trans
test.rs:1 use std::marker::{MarkerTrait, PhantomFn};
          ^
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /Users/rustbuild/src/rust-buildbot/slave/beta-dist-rustc-mac/build/src/libsyntax/diagnostic.rs:149
```
```
~/Downloads $ rustc --version
rustc 1.0.0-beta.2 (e9080ec39 2015-04-16) (built 2015-04-16)
```

**Comment by arielb1** [comment]
Example uses an unconstrained projection (`ToPtr<*mut <PAR as Wrapper>::Inner> for CHI`) - fixed by rust-lang/rfcs#447. Closing.
